src/fileio.cpp

```diff
@@ -51,85 +48,99 @@ const char* LoadingError::what() const noexcept
 }
 
 /**
- * @fn void DirectoryReader::OpenPath(const char *path)
- * Set up the directory reader object for reading a directory.
- * @param path Path to the directory.
- * @note Failure is not reported here, but #NextEntry will not return anything useful.
+ * Scan a given directory and return all entries.
+ * @param Directory to scan.
  */
+std::vector<std::string> GetAllEntries(const std::string &path)
+{
+	std::vector<std::string> entries;
+	if(!PathIsDirectory(path)) return entries;
 
-/**
- * @fn const char *DirectoryReader::NextEntry()
- * Get next entry of the directory contents.
- * @return Pointer to name of next entry (as a file path suitable for opening a file). Returns \c nullptr if not next entry exists.
- * @note The memory returned is owned by the #DirectoryReader object, and should not be released.
- */
+	for(const auto &entry : std::filesystem::directory_iterator(path)) {
+		std::string entryName = entry.path().string();
+		entries.push_back(entryName);
+	}
+	return entries;
+}
 
 /**
- * @fn void DirectoryReader::ClosePath()
- * Denote no further interest in reading the contents of the current directory.
+ * Scan a given directory and return all file entries.
+ * @param Directory to scan.
  */
+std::vector<std::string> GetAllFileEntries(const std::string &path)
+{
+	std::vector<std::string> entries;
+	if(!PathIsDirectory(path)) return entries;
 
-/**
- * @fn const char *DirectoryReader::MakePath(const char *directory, const char *fname)
- * Construct a path from a directory and a file name.
- * @param directory Directory part of the path.
- * @param fname File name part of the path.
- * @return The combined path to the file.
- */
+	for(const auto &entry : std::filesystem::directory_iterator(path)) {
+		if(entry.is_regular_file()) {
+			std::string entryName = entry.path().string();
+			entries.push_back(entryName);
+		}
+	}
+	return entries;
+}
 
-/**
- * @fn bool DirectoryReader::EntryIsFile()
- * Test whether the last returned entry from #NextEntry is a file.
- * @return Whether the entry is a file.
- */
 
 /**
- * @fn bool DirectoryReader::EntryIsDirectory()
- * Test whether the last returned entry from #NextEntry is a directory.
- * @return Whether the entry is a directory.
+ * Delete the given file.
+ * @param path Path to delete.
  */
+void RemoveFile(const std::string &path)
+{
+
+	if (!PathIsFile(path)) return;
+	try {
+		std::filesystem::remove(path);
+	} catch(std::filesystem::filesystem_error const &ex) {
+		error("%s", ex.what());
```

Comment:
I'm not sure about catching the error and just printing it here. I think either you should let the caller handle the error (more complicated) or crash the program (easier). This pattern is repeated in some other places.

Reply:
Hmm maybe in void functions we can let the program crash and in  non-void function we return some error value (like false in boolean functions)

Reply:
Another options could be to let the exception propagate instead of catching it. Another option (pun unintended) would be to use `std::optional`: https://en.cppreference.com/w/cpp/utility/optional

Reply:
Propagating seems a better solution to me